joern-cli/frontends/csharpsrc2cpg/src/main/scala/io/joern/csharpsrc2cpg/utils/DependencyDownloader.scala

```diff
@@ -0,0 +1,186 @@
+package io.joern.csharpsrc2cpg.utils
+
+import better.files.File
+import io.joern.csharpsrc2cpg.Config
+import io.joern.csharpsrc2cpg.datastructures.CSharpProgramSummary
+import io.joern.x2cpg.astgen.AstGenRunner.DefaultAstGenRunnerResult
+import io.shiftleft.codepropertygraph.generated.Cpg
+import io.shiftleft.codepropertygraph.generated.nodes.Dependency
+import io.shiftleft.semanticcpg.language.*
+import org.slf4j.LoggerFactory
+
+import java.io.FileOutputStream
+import java.net.{HttpURLConnection, URI, URLConnection}
+import java.util.zip.{GZIPInputStream, InflaterInputStream, ZipEntry}
+import scala.util.{Failure, Success, Try, Using}
+
+/** Queries NuGet for the artifacts of a programs' dependencies, according to the V2 API.
+  *
+  * TODO: Note that NuGet has API throttling, so beware of that if this is to be processed concurrently.
+  *
+  * @see
+  *   <a href="https://learn.microsoft.com/en-us/nuget/api/overview">NuGet API</a>
+  */
+class DependencyDownloader(
+  cpg: Cpg,
+  config: Config,
+  internalProgramSummary: CSharpProgramSummary,
+  nugetRepositoryApi: String = "www.nuget.org/api/v2"
+) {
+
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  /** Downloads dependencies and summarises their symbol information.
+    * @return
+    *   the dependencies' summary combined with the given internal program summary.
+    */
+  def download(): CSharpProgramSummary = {
+    File.temporaryDirectory("joern-csharpsrc2cpg").apply { dir =>
+      cpg.dependency.filterNot(isAlreadySummarized).foreach(downloadDependency(dir, _))
+      unzipDependencies(dir)
+      summarizeDependencies(dir) ++ internalProgramSummary
+    }
+  }
+
+  /** Using the internal program summary (which may contain pre-defined type stubs), determines if the dependency is
+    * already summarized and does not need to be downloaded.
+    * @param dependency
+    *   the dependency to download.
+    * @return
+    *   true if the dependency is already in the given summary, false if otherwise.
+    */
+  private def isAlreadySummarized(dependency: Dependency): Boolean = {
+    // TODO: Implement
+    false
+  }
+
+  /** Downloads the dependency to a temporary directory. This will be a `nupkg` and `snupkg` file for each dependency
+    * respectively, which are compressed using the ZIP format.
+    *
+    * @param targetDir
+    *   the directory to download to.
+    * @param dependency
+    *   the dependency to download.
+    * @return
+    *   a disposable version of the directory where the dependencies live.
+    */
+  private def downloadDependency(targetDir: File, dependency: Dependency): Unit = {
+    Seq("package", "symbolpackage")
+      .map(packageType =>
+        URI(s"https://$nugetRepositoryApi/$packageType/${dependency.name}/${dependency.version}").toURL
```

Comment:
In relation to https://github.com/joernio/joern/pull/4295 we might get no versions - this is actually a common case I am seeing (eg. https://github.com/dotnet-architecture/eShopOnWeb/blob/main/src/ApplicationCore/ApplicationCore.csproj)  When this happens, the `package` gets downloaded fine because NuGet API redirects to latest package. However we get a 400 since `symbolpackage` expects exact `$dependency.version`. Here is an example: 
 
400 : https://www.nuget.org/api/v2/symbolpackage/System.IdentityModel.Tokens.Jwt/
200: https://www.nuget.org/api/v2/symbolpackage/System.IdentityModel.Tokens.Jwt/7.4.0

One workaround could be to first download the nupkg, get it's version (which is latest) and then append to the `symbolpackage` URL and re-download.

A better approach could be to do something like this: https://api.nuget.org/v3-flatcontainer/System.IdentityModel.Tokens.Jwt/index.json and fetch package versions and then pick last one (Ref: https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions) Note, this is using v3 API and to me seems a more robust way.